File: laptop-app/src/output/TerminalOutputProcessor.ts
/**
 * TerminalOutputProcessor
 * -----------------------
 * Mirrors the Swift implementation that keeps track of the previously
 * processed screen output and extracts only the newly added lines.
 */

export class TerminalOutputProcessor {
  private lastProcessedScreenOutput = '';

  extractNewLines(currentOutput: string): string {
    if (this.lastProcessedScreenOutput.length === 0) {
      this.lastProcessedScreenOutput = currentOutput;
      return currentOutput;
    }

    const currentCleaned = this.removeAnsiCodes(currentOutput);
    const lastCleaned = this.removeAnsiCodes(this.lastProcessedScreenOutput);

    if (currentCleaned === lastCleaned) {
      return '';
    }

    if (currentCleaned.startsWith(lastCleaned)) {
      const newContent = currentCleaned.slice(lastCleaned.length);
      this.lastProcessedScreenOutput = currentOutput;
      return newContent.trim().length === 0 ? '' : newContent;
    }

    if (currentCleaned.endsWith(lastCleaned) && currentCleaned.length > lastCleaned.length) {
      const newContent = currentCleaned.slice(0, currentCleaned.length - lastCleaned.length);
      this.lastProcessedScreenOutput = currentOutput;
      return newContent.trim().length === 0 ? '' : newContent;
    }

    const currentLines = currentCleaned.split('\n');
    const lastLines = lastCleaned.split('\n');

    let newLinesStartIndex = 0;
    const minLength = Math.min(currentLines.length, lastLines.length);

    if (currentLines.length > lastLines.length) {
      newLinesStartIndex = lastLines.length;
    } else {
      for (let i = minLength - 1; i >= 0; i -= 1) {
        const currentLine = currentLines[i].trim();
        const lastLine = lastLines[i].trim();

        if (currentLine !== lastLine) {
          newLinesStartIndex = i;
          break;
        }
      }
    }

    if (newLinesStartIndex < currentLines.length) {
      const newLines = currentLines.slice(newLinesStartIndex);
      const result = newLines.join('\n');
      this.lastProcessedScreenOutput = currentOutput;
      return result.trim().length === 0 ? '' : result;
    }

    this.lastProcessedScreenOutput = currentOutput;
    return '';
  }

  reset(): void {
    this.lastProcessedScreenOutput = '';
  }

  private removeAnsiCodes(text: string): string {
    const pattern = /\u001B\[[0-9;]*[a-zA-Z]/g;
    return text.replace(pattern, '');
  }
}

File: EchoShell/EchoShell/Services/TerminalOutputProcessor.swift
//
//  TerminalOutputProcessor.swift
//  EchoShell
//
//  Processes terminal output and extracts meaningful content
//

import Foundation

class TerminalOutputProcessor {
    private var lastProcessedScreenOutput: String = ""
    
    // Extract only new lines from screen output
    func extractNewLines(from currentOutput: String) -> String {
        if lastProcessedScreenOutput.isEmpty {
            lastProcessedScreenOutput = currentOutput
            return currentOutput
        }
        
        let currentCleaned = removeAnsiCodes(from: currentOutput)
        let lastCleaned = removeAnsiCodes(from: lastProcessedScreenOutput)
        
        if currentCleaned == lastCleaned {
            return ""
        }
        
        // Check if new content was appended
        if currentCleaned.hasPrefix(lastCleaned) {
            let newContent = String(currentCleaned.dropFirst(lastCleaned.count))
            lastProcessedScreenOutput = currentOutput
            return newContent.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? "" : newContent
        }
        
        // Check if content was prepended
        if currentCleaned.hasSuffix(lastCleaned) && currentCleaned.count > lastCleaned.count {
            let newContent = String(currentCleaned.dropLast(lastCleaned.count))
            lastProcessedScreenOutput = currentOutput
            return newContent.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? "" : newContent
        }
        
        // Compare line by line
        let currentLines = currentCleaned.components(separatedBy: .newlines)
        let lastLines = lastCleaned.components(separatedBy: .newlines)
        
        var newLinesStartIndex = 0
        let minLength = min(currentLines.count, lastLines.count)
        
        if currentLines.count > lastLines.count {
            newLinesStartIndex = lastLines.count
        } else {
            for i in stride(from: minLength - 1, through: 0, by: -1) {
                let currentLine = currentLines[i].trimmingCharacters(in: .whitespaces)
                let lastLine = lastLines[i].trimmingCharacters(in: .whitespaces)
                
                if currentLine != lastLine {
                    newLinesStartIndex = i
                    break
                }
            }
        }
        
        if newLinesStartIndex < currentLines.count {
            let newLines = Array(currentLines[newLinesStartIndex...])
            let result = newLines.joined(separator: "\n")
            lastProcessedScreenOutput = currentOutput
            return result.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? "" : result
        }
        
        lastProcessedScreenOutput = currentOutput
        return ""
    }
    
    func reset() {
        lastProcessedScreenOutput = ""
    }
    
    // Helper function to remove ANSI codes
    private func removeAnsiCodes(from text: String) -> String {
        let pattern = "\\u{001B}\\[[0-9;]*[a-zA-Z]"
        guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else {
            return text
        }
        let range = NSRange(text.startIndex..., in: text)
        return regex.stringByReplacingMatches(in: text, options: [], range: range, withTemplate: "")
    }
}

